In [1]:
%%writefile /tmp/gen_cache.py
import json

PAGES = 16          # one declared address per 4 KiB page -> a 64 KiB footprint
REPEAT = 8          # reload each address, so the timed window is long enough

lines = []
for _ in range(REPEAT):
    for page in range(PAGES):
        lines.append(f"    x += arr[{page * 4096}];")

source = f"""
extern "C" long cache_loads(const unsigned char* arr) {{
    long x = 0;
{chr(10).join(lines)}
    return x;
}}
"""

BASE = 0x10000000

# One byte declared per page, not a whole page. perf tracks every declared
# address, and a dense 64 KiB payload asks for a 192 GiB input buffer.
PAYLOAD = {
    hex(BASE + page * 4096): [(page * 37 + 13) % 256] for page in range(PAGES)
}

with open("/tmp/study_cache.cpp", "w") as fh:
    fh.write(source)

with open("/tmp/study_cache_data.json", "w") as fh:
    json.dump({"base": BASE, "payload": PAYLOAD, "pages": PAGES}, fh)

Overwriting /tmp/gen_cache.py


In [2]:
!python /tmp/gen_cache.py && head -6 /tmp/study_cache.cpp


extern "C" long cache_loads(const unsigned char* arr) {
    long x = 0;
    x += arr[0];
    x += arr[4096];
    x += arr[8192];


In [3]:
!g++ -O3 -c /tmp/study_cache.cpp -o /tmp/study_cache.o

In [4]:
import json
import subprocess

BASE = 0x10000000
_pages = 16

text = subprocess.run(
    ["objdump", "-d", "--no-show-raw-insn", "/tmp/study_cache.o"],
    capture_output=True, text=True, check=True,
).stdout
body = text.split("<cache_loads>:", 1)[1]
insns = [l for l in body.splitlines() if l.strip() and l.strip()[0].isdigit()]
loads = sum(1 for l in insns if "(%rdi)" in l or "(rdi)" in l)
print(f"{len(insns)} instructions, {loads} loads, footprint {_pages * 4096 // 1024} KiB")

153 instructions, 11 loads, footprint 64 KiB


## Backend bound: hot vs cold data cache

One axis, one function, two tiers. `cache_loads` makes ~128 byte loads spread
across a 64 KiB footprint — one declared address per 4 KiB page, so every load
lands on its own cache line.

Everything about the instruction stream is identical between the two runs. The
only difference is where perf says the data lives: `dcache: hot` keeps the lines
resident, `dcache: cold` evicts them before each call with `mprotect`, so the
loads have to come from DRAM.

**Claim.** With `dcache` set to `cold`, the same instructions cost far more
cycles, retire at a much lower IPC, and `topdown-be-bound` becomes the dominant
slot.

In [5]:
import json

import numpy as np
import pandas as pd

import perf

TARGET = "/tmp/study_cache.o"
FUNC = "cache_loads"

_meta = json.load(open("/tmp/study_cache_data.json"))
BASE = _meta["base"]
PAGES = _meta["pages"]
# perf takes the base pointer in the usual first-argument register, and the
# payload as address -> bytes. One byte per page keeps the tracked-address count
# at 16; declaring a dense 64 KiB buffer instead asks for a 192 GiB input buffer.
DATA = {"arg0": BASE, **_meta["payload"]}

SLOTS = [
    "topdown-retiring",
    "topdown-bad-spec",
    "topdown-fe-bound",
    "topdown-be-bound",
]
RAW = ["cycles", "instructions", "cache-references", "cache-misses"]
TIERS = ["hot", "cold"]

CONFIG = {
    "samples": 20,
    "iterations": 100,
    "branch": "predictable",
    "dtlb": "hot",
    "itlb": "hot",
    "icache": "hot",
    "code": {"align": 16},
    "func": {"align": 16, "order": "as-is"},
    "stack": {"size": 0x200000, "align": 16},
    "backend": {"loop": {"runs": 4}},
}


def sweep(event):
    frames = []
    for tier in TIERS:
        df = perf.benchmark(
            code=[TARGET, FUNC],
            mode=["latency"],
            event=event,
            config={**CONFIG, "dcache": tier},
            data=DATA,
        ).reset_index()
        df["dcache"] = tier
        frames.append(df)
    return pd.concat(frames, ignore_index=True)


def usable(series):
    """Miss counters are zero-inflated and arrive with NaN; see the note below."""
    values = series.to_numpy(dtype=float)
    return values[~np.isnan(values)]


def median_or_zero(series):
    kept = usable(series)
    return float(np.median(kept)) if kept.size else 0.0


def check(claim, ok):
    print(("PASS  " if ok else "FAIL  ") + claim)
    assert ok, claim

In [6]:
raw_top = sweep(SLOTS)
print("raw slot counts (fe-bound reads NaN when the core never stalls the "
      "front end, so it is filled with 0 for the shares):")
raw_med = raw_top.groupby("dcache")[SLOTS].median()
print(raw_med.round(1).to_string())
print()

TOP = raw_med.fillna(0.0)
share = TOP.div(TOP.sum(axis=1), axis=0).round(3)
print(share.to_string())
print()

RAWDF = sweep(RAW)
summary = pd.DataFrame(
    {
        "cycles": RAWDF.groupby("dcache")["cycles"].median(),
        "instructions": RAWDF.groupby("dcache")["instructions"].median(),
        "cache_misses": RAWDF.groupby("dcache")["cache-misses"].apply(median_or_zero),
    }
)
summary["ipc"] = summary["instructions"] / summary["cycles"]
print(summary.round(2).to_string())
print()

hot_cyc = summary.loc["hot", "cycles"]
cold_cyc = summary.loc["cold", "cycles"]
hot_insn = summary.loc["hot", "instructions"]
cold_insn = summary.loc["cold", "instructions"]
hot_be = share.loc["hot", "topdown-be-bound"]
cold_be = share.loc["cold", "topdown-be-bound"]
fe_share = share["topdown-fe-bound"]

check(
    f"both tiers retire the same instructions "
    f"({hot_insn:.0f} hot vs {cold_insn:.0f} cold)",
    0.99 <= cold_insn / hot_insn <= 1.01,
)
check(
    f"cold costs much more cycles ({cold_cyc:.0f} vs {hot_cyc:.0f}, "
    f"{cold_cyc / hot_cyc:.1f}x)",
    cold_cyc > 3 * hot_cyc,
)
check(
    f"IPC collapses ({summary.loc['hot', 'ipc']:.2f} -> "
    f"{summary.loc['cold', 'ipc']:.2f})",
    summary.loc["cold", "ipc"] < 0.5 * summary.loc["hot", "ipc"],
)
check(
    f"backend-bound becomes the dominant slot ({hot_be} hot -> {cold_be} cold)",
    cold_be > 0.8 and cold_be > hot_be,
)
check(
    f"the front end is not the story either "
    f"(fe-bound share {fe_share.max():.2f})",
    fe_share.max() < 0.1,
)
check(
    f"and the miss counter agrees ({summary.loc['hot', 'cache_misses']:.0f} -> "
    f"{summary.loc['cold', 'cache_misses']:.0f}), so this really is a data miss",
    summary.loc["cold", "cache_misses"] > summary.loc["hot", "cache_misses"],
)

raw slot counts (fe-bound reads NaN when the core never stalls the front end, so it is filled with 0 for the shares):
        topdown-retiring  topdown-bad-spec  topdown-fe-bound  topdown-be-bound
dcache                                                                        
cold               184.0             222.5               5.0            3536.0
hot                184.0             220.0               5.0             431.0

        topdown-retiring  topdown-bad-spec  topdown-fe-bound  topdown-be-bound
dcache                                                                        
cold               0.047             0.056             0.001             0.896
hot                0.219             0.262             0.006             0.513



        cycles  instructions  cache_misses   ipc
dcache                                          
cold    1026.0         183.0          16.0  0.18
hot      146.0         183.0           0.0  1.25

PASS  both tiers retire the same instructions (183 hot vs 183 cold)
PASS  cold costs much more cycles (1026 vs 146, 7.0x)
PASS  IPC collapses (1.25 -> 0.18)
PASS  backend-bound becomes the dominant slot (0.513 hot -> 0.896 cold)
PASS  the front end is not the story either (fe-bound share 0.01)
PASS  and the miss counter agrees (0 -> 16), so this really is a data miss


In [7]:
perf.plot(TOP, "boxen", event="topdown-be-bound", groupby="dcache")


P0;0;0q"1;1;637;485#0;2;0;0;0#1;2;4;4;4#2;2;8;8;8#3;2;11;11;11#4;2;14;14;14#5;2;17;17;17#6;2;21;21;21#7;2;24;24;24#8;2;26;26;26#9;2;29;29;27#10;2;27;28;29#11;2;30;30;30#12;2;33;33;33#13;2;36;36;36#14;2;39;39;39#15;2;42;42;42#16;2;46;46;46#17;2;49;49;49#18;2;74;52;13#19;2;75;53;13#20;2;76;53;13#21;2;9;42;61#22;2;9;42;60#23;2;52;52;52#24;2;55;55;55#25;2;58;58;58#26;2;61;61;61#27;2;64;64;64#28;2;67;67;67#29;2;70;70;70#30;2;74;74;74#31;2;77;77;77#32;2;80;80;80#33;2;83;83;83#34;2;86;86;86#35;2;89;89;89#36;2;93;93;93#37;2;95;95;95#38;2;100;100;100#38!637~-!637~-!564~N^^N^^N!7^N^^N!4^N!22~FFF!6~FFF!16~$#37!564?O??O??O!7?O??O!4?O!22?G?G!6?G#34?o$#10!564?_#6!4_?___?!12_#10_#33!22?O#0o!8?o#36G$#7!569?_#5???_#34!35?_#24G!8?G$#35!611?o!6?o-#38!563~v#6~#21}!18~{C#6~!6?G#1A!7?A??A#30A!8?@?@#6A$#37!563?G#22?@!18?Bz#38?v!4~B@w{{wx~B@w{{?@F~???~B@w{{???!16~$#37!587?G#15!4?G!7?_A#25?@#34?G???G!8?O?}$#11!592?O??A#27@!7?@#35_!5?~!6?@$#14!592?_#0C!7?C!4?w???~???C!4?~$#30!592?C#8_!7?_??A!9?g#2A#26@#7A$#12

[]

In [8]:
perf.plot(RAWDF, "boxen", event="cycles", groupby="dcache")


P0;0;0q"1;1;637;485#0;2;0;0;0#1;2;5;5;5#2;2;8;8;8#3;2;11;11;11#4;2;14;14;14#5;2;17;17;17#6;2;21;21;21#7;2;23;23;23#8;2;25;25;24#9;2;27;27;27#10;2;29;29;27#11;2;27;28;29#12;2;30;30;30#13;2;33;33;33#14;2;36;36;36#15;2;39;39;39#16;2;42;42;42#17;2;45;45;45#18;2;48;48;48#19;2;68;48;14#20;2;74;52;13#21;2;75;53;13#22;2;76;53;13#23;2;9;42;61#24;2;9;42;60#25;2;51;51;51#26;2;55;55;55#27;2;58;58;58#28;2;61;61;61#29;2;64;64;64#30;2;67;67;67#31;2;71;71;71#32;2;74;74;74#33;2;77;77;77#34;2;80;80;80#35;2;83;83;83#36;2;86;86;86#37;2;89;89;89#38;2;93;93;93#39;2;95;95;95#40;2;100;100;100#40!637~-!637~-!564~N^^N^^N^^N!4^N^^N!4^N!5~FFF!15~NNN!24~$#39!564?O??O??O??O!4?O??O!4?O!7?G#38!15?O?_$#11!564?_#6___??!13_?__#11_#38!5?G?o#35!15?_#0_#39O$#7!568?__!13?_#34!8?o#0o#26!17?O$#27!593?G-#40!563~v#6~#23}!15~}~~}{#6~!7?A#9A!6?_??A#37_!5?k#26@#28A$#39!563?G#24?@!15?@??@B#40?Z!4~???{{?@B~B@w{{?@F{???{{!22~$#39!587?c!6?`??o!8?O??B#15??A$#34!592?~!6?g???D!4?_#25???@$#0!593?~!4?{???C!4?w???~$#38!594?O??G?C!6?@!5?O$

[]

## The payload mistake, and the NaN counter

**A dense 64 KiB payload asks for a 192 GiB buffer.** The first version of this
notebook declared a full 64 KiB array as one contiguous
`hex(BASE) + ":": [...]` blob. perf tracks every declared address, so the input
buffer it builds came out at 192 GiB and the cell died on allocation rather than
returning a wrong number. Declaring **one byte per 4 KiB page** gives the same
footprint with 16 tracked addresses and a buffer of a few kilobytes.

**`cache-misses` is zero-inflated and arrives with `NaN`.** Under `hot` almost
every sample misses zero times, and those samples read back as `NaN` rather than
`0`. A plain `.median()` of that column therefore returns `NaN` — which is why the
earlier version of this notebook printed `cache-misses NaN` and could not assert
anything about it. `median_or_zero` drops the `NaN`s first. The
bad-speculation notebook hit the same thing with `branch-misses`, so it is a
property of the miss counters rather than of this axis.

**Why `hot` still shows a bad-spec share.** Under `hot` the loads hit L1 and the
function retires fast, so the harness' own loop and call overhead becomes a
visible fraction of a 146-cycle window — `topdown-bad-spec` reads 0.279 there.
That share is the measurement scaffolding, not the function. The cold case runs
989 cycles, where the same overhead is 2%, and the slots describe the memory
stalls.